### ISSUE TITLE
Can not implicitly convert non-nullable generic argument to nullable ones

### ISSUE THREAD

**Issue by dkwingsmt** [dart-lang/language#1572]
In the following snippet, we try to process a nullable list, so that if the list _is_ null, we instead get a list _with_ null. (The resulting `Iterable` would actually be used in a for-loop.)

We would like to write it in `testShouldPass`, but the compiler (analyzer?) complains "A value of type 'Object' can't be assigned to a variable of type 'Iterable<Child?>'.". We have to instead use the code in `workaround`.

```dart
class Child {}

class Family {
  Iterable<Child>? get children => const Iterable<Child>.empty();
}

class Test {
  static void testShouldPass(Family family) {
    final Iterable<Child?> maybeChildren = family.children ?? <Child?>[null];
  }

  static void workaround(Family family) {
    final Iterable<Child?>? maybeEmptyChildren = family.children;
    final Iterable<Child?> maybeChildren = maybeEmptyChildren ?? <Child?>[null];
  }
}
```

Version: I'm currently on 2.13.0-222.0-dev, and it also occurred in 2.10

**Comment by srawlins** [comment]
The type of `family.children ?? <Child?>[null]` is the Least Upper Bound (LUB) of the type on the left (after promoting to non-null) and the type on the right. The type on the left is the non-nullable-promoted type of `family.children`, which is `Iterable<Child>`. The type on the right is the type of `<Child?>[]`, which is `List<Child?>`. The LUB of `Iterable<Child>` and `List<Child?>` is unfortunately `Object` (I think).

In the LUB calculation, it gets muddy having `A<C?>` and `B<C>`. The code works out if you cast `family.children` to have the same type argument as the right side (`Child?`):

```dart
final Iterable<Child?> maybeChildren = family.children?.cast<Child?>() ?? <Child?>[null];
```

**Comment by dkwingsmt** [comment]
Yeah, it's just a little frustrating that something works out without any casts but only an intermediate variable. If I ever encounter this again, I'll have to comment why this seemingly unnecessary variable is needed.

Thanks for the answer. Feel free to close if this does not feel like an issue.
